x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/data/MockBlockFactory.java

```diff
@@ -0,0 +1,366 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.compute.data;
+
+import org.apache.lucene.tests.util.LuceneTestCase;
+import org.apache.lucene.util.BytesRef;
+import org.elasticsearch.common.breaker.CircuitBreaker;
+import org.elasticsearch.common.util.BigArrays;
+import org.elasticsearch.common.util.BytesRefArray;
+import org.elasticsearch.compute.data.Block.MvOrdering;
+
+import java.util.BitSet;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+public class MockBlockFactory extends BlockFactory {
+
+    static final boolean TRACK_ALLOCATIONS = true;
+
+    static Object trackDetail() {
+        return TRACK_ALLOCATIONS
+            ? new RuntimeException("Block allocated from test: " + LuceneTestCase.getTestClass().getName()) // TODO: message
+            : true;
+    }
+
+    final ConcurrentMap<Object, Object> TRACKED_BLOCKS = new ConcurrentHashMap<>();
+
+    public void ensureAllBlocksAreReleased() {
+        purgeTrackBlocks();
+        final Map<Object, Object> copy = new HashMap<>(TRACKED_BLOCKS);
+        // we should really assert this, but not just yet, see comment below
+        // assert breaker().getUsed() > 0 : "Expected some used in breaker if tracked blocks is not empty";
+        if (breaker().getUsed() == 0) {
+            TRACKED_BLOCKS.clear();
+            return; // this is a hack until we get better a mock tracking.
+        }
+        if (copy.isEmpty() == false) {
+            Iterator<Object> causes = copy.values().iterator();
+            Object firstCause = causes.next();
+            RuntimeException exception = new RuntimeException(
+                copy.size() + " blocks have not been released",
+                firstCause instanceof Throwable ? (Throwable) firstCause : null
+            );
+            while (causes.hasNext()) {
+                Object cause = causes.next();
+                if (cause instanceof Throwable) {
+                    exception.addSuppressed((Throwable) cause);
+                }
+            }
+            throw exception;
+        }
+    }
+
+    public MockBlockFactory(CircuitBreaker breaker, BigArrays bigArrays) {
+        super(breaker, bigArrays);
+    }
+
+    @Override
+    void adjustBreaker(final long delta, final boolean isDataAlreadyCreated) {
+        purgeTrackBlocks();
+        super.adjustBreaker(delta, isDataAlreadyCreated);
+    }
+
+    void purgeTrackBlocks() {
+        for (var entries : TRACKED_BLOCKS.entrySet()) {
+            var b = entries.getKey();
+            if (b instanceof Block block) {
```

Comment:
Oh that's a nightmare.

Reply:
We should make this less scary. But it can wait.

Reply:
yeah. we'll eventually improve this and the blocks, vectors, and builders that it inspects.